runtime/sema/check_composite_declaration.go

```diff
@@ -1977,6 +1996,120 @@ func (checker *Checker) enumMembersAndOrigins(
+		identifier := arg.Identifier.Identifier
+		// these are guaranteed to exist at time of destruction, so we allow them
+		if identifier == SelfIdentifier || identifier == BaseIdentifier {
+			break
+		}
+		// if it's an attachment, then it's also okay
+		if checker.typeActivations.Find(identifier) != nil {
+			break
+		}
+		checker.report(&DefaultDestroyInvalidArgumentError{
+			Range: ast.NewRangeFromPositioned(checker.memoryGauge, arg),
+		})
+	case *ast.MemberExpression:
+		checker.checkDefaultDestroyParamExpressionKind(arg.Expression)
+	case *ast.IndexExpression:
+		checker.checkDefaultDestroyParamExpressionKind(arg.TargetExpression)
+		checker.checkDefaultDestroyParamExpressionKind(arg.IndexingExpression)
+
+		// indexing expressions on arrays can fail, and must be disallowed, but
+		// indexing expressions on dicts, or composites (for attachments) will return `nil` and thus never fail
+		targetExprType := checker.Elaboration.IndexExpressionTypes(arg).IndexedType
+		// `nil` indicates that the index is a type-index (i.e. for an attachment access)
+		if targetExprType == nil {
+			return
+		}
+
+		switch targetExprType := targetExprType.(type) {
+		case *DictionaryType:
+			return
+		case *ReferenceType:
+			if _, isDictionaryType := targetExprType.Type.(*DictionaryType); isDictionaryType {
+				return
+			}
+		}
+
+		checker.report(&DefaultDestroyInvalidArgumentError{
+			Range: ast.NewRangeFromPositioned(checker.memoryGauge, arg),
+		})
+
+	default:
+		checker.report(&DefaultDestroyInvalidArgumentError{
+			Range: ast.NewRangeFromPositioned(checker.memoryGauge, arg),
+		})
+	}
+}
+
+func (checker *Checker) checkDefaultDestroyEventParam(
+	param Parameter,
+	astParam *ast.Parameter,
+	containerType ContainerType,
+	containerDeclaration ast.Declaration,
+) {
+	paramType := param.TypeAnnotation.Type
+	paramDefaultArgument := astParam.DefaultArgument
+
+	// make `self` and `base` available when checking default arguments so the fields of the composite are available
+	checker.declareSelfValue(containerType, containerDeclaration.DeclarationDocString())
+	if compositeContainer, isComposite := containerType.(*CompositeType); isComposite && compositeContainer.Kind == common.CompositeKindAttachment {
+		checker.declareBaseValue(
+			compositeContainer.baseType,
+			compositeContainer,
+			compositeContainer.baseTypeDocString)
+	}
+	param.DefaultArgument = checker.VisitExpression(paramDefaultArgument, paramType)
+
+	unwrappedParamType := UnwrapOptionalType(paramType)
+	// default events must have default arguments for all their parameters; this is enforced in the parser
+	// we want to check that these arguments are all either literals or field accesses, and have primitive types
+	if !IsSubType(unwrappedParamType, StringType) &&
+		!IsSubType(unwrappedParamType, NumberType) &&
+		!IsSubType(unwrappedParamType, TheAddressType) &&
+		!IsSubType(unwrappedParamType, PathType) &&
+		!IsSubType(unwrappedParamType, BoolType) {
```

Comment:
This set of types seems to need to be "in sync" / is similar to the set of expressions in `checkDefaultDestroyParamExpression`. Maybe there's a way to make this a bit easier to maintain?

Reply:
Yeah, we just don't really have a good definition of what a "primitive" type is in Cadence right now, and no good way to really check that statically in Go. If we can produce a concrete definition (perhaps in a subsequent PR) then that would be a good change her.e

Reply:
Could you please open a tech debt issue for that?

Reply:
https://github.com/onflow/cadence/issues/2929